torch/_subclasses/meta_utils.py

```diff
@@ -269,6 +269,180 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
+        #     is fully fake with no invalid specialization.
+        def view_from_base(base, t, source=source, shape_env=shape_env):
+            # fake-ify t's metadata according to the outer symbolic context
+            (sizes, strides, storage_offset) = sym_sizes_strides_storage_offset(
+                t, source
+            )
+            if not is_traceable_wrapper_subclass(
+                t
+            ) and not is_traceable_wrapper_subclass(base):
+                # Dense -> Dense view case uses as_strided() to construct view relationship.
+                # TODO: Change this logic to use view replay for consistency?
+                # It's likely there is no view func available.
+                return base.as_strided(sizes, strides, storage_offset)
+
+            from torch._dynamo.source import EphemeralSource
+
+            def symint_visitor_fn(s):
+                if shape_env is None:
+                    return s
+
+                # NB: The symbol here is expected to be simplified out.
+                sym_source = EphemeralSource("symint_visitor_fn")
+                symbol = shape_env.create_symbol(s, sym_source)
+                return shape_env.create_symintnode(symbol, hint=s, source=sym_source)
+
+            real_to_fake_mapping = {}
+            if is_traceable_wrapper_subclass(t):
+                # Fake-ify t naively here; this is only done so we can get fake-ified inner
+                # tensors with the correct relationships to the outer sizes / strides for use
+                # in view replay. It's done beforehand here because it's not easy to do when
+                # visiting tensors one-by-one during view replay.
+                #
+                # Example:
+                #   Consider a Dense -> NJT view. NJT has (values, offsets) components and we
+                #   want a view of values with the offsets closed over. As the offsets component
+                #   is needed to describe the output view, it's important that it's fakeified
+                #   correctly.
+                fake_t = empty_create_subclass(
+                    t, outer_size=sizes, outer_stride=strides
+                )
+                attrs, _ = fake_t.__tensor_flatten__()
+                for attr in attrs:
+                    real_to_fake_mapping[getattr(t, attr)] = getattr(fake_t, attr)
+
+            def tensor_visitor_fn(t):
+                # Fake inner tensors of view subclasses will come from the mapping built above.
+                fake_t = real_to_fake_mapping.get(t, None)
+                if fake_t is not None:
+                    return fake_t
+
+                # For other closed-over tensor state, fake-ify it as all dynamic with an
+                # ephemeral source. This avoids invalid specialization during view replay.
+                return all_dynamic_meta_tensor(
+                    t, source=EphemeralSource("tensor_visitor_fn")
```

Comment:
I was thinking about how this relates to our current handling for dense-dense views. Currently, we generate an exotic `x._base` source for the SymInts that arise here, but we then suppress guard creation when running the view function, so as to avoid actually guarding on base. And then we sort of blithely pretend that we are base agnostic.

What we are doing here is a bit better: we're trying hard to eliminate the inner SymInts so that we don't have guards referencing them. But if this still is not working, we may want to just apply the same suppress guards playbook. (That being said, the whole dense view situation is quite unsatisfactory and I'm not sure what to do about it.)

Reply:
added a comment here about the option to suppress guards